File: smeshram1/itmd-521/labs/week-04/py/assignment_01.py
import sys
from pyspark.sql import SparkSession
from pyspark.sql.types import *

if __name__ == "_main_":
    if len(sys.argv) <= 0:
        sys.exit(1)

    spark = SparkSession.builder.appName("Divvy_Trip").getOrCreate()
    # Source file
    csvFile = sys.argv[1]
    
    # Infer the Schema
    inferDataFrame = spark.read.format("csv").option("header", "true").option("inferSchema", "true").load(csvFile)
    print("**************************************")
    inferDataFrame.show(n=10)
    print("***************Schema inferred **************")
    inferDataFrame.printSchema()
    print("**************Number of records*****************")        
    print("Number of records in infer DataFrame = " + str(inferDataFrame.count()))
    print("****************End of infer schema***************")
    
    # Defining schema programmatically 
    schema = StructType([
        StructField("trip_id", IntegerType()),
        StructField("starttime", StringType()),
        StructField("stoptime", StringType()),
        StructField("bikeid", IntegerType()),
        StructField("tripduration", IntegerType()),
        StructField("from_station_id", IntegerType()),
        StructField("from_station_name", StringType()),
        StructField("to_station_id", IntegerType()),
        StructField("to_station_name", StringType()),
        StructField("usertype", StringType()),
        StructField("gender", StringType()),
        StructField("birthyear", IntegerType())
    ])
    
    structDataFrame = spark.read.schema(schema).csv(csvFile)
    print("**************************************")
    structDataFrame.show(n=10)
    print("************Print Schema Programmatically*************")
    structDataFrame.printSchema()
    print("**************Number of records*******************")
    print("Number of records in structure DataFrame = " + str(structDataFrame.count()))
    print("****************End of structured schema***************")

    
    # Attaching a schema via DDL
    ddlSchema = "trip_id INT,starttime STRING,stoptime STRING,bikeid INT,tripduration INT,from_station_id INT,from_station_name STRING,to_station_id INT,to_station_name STRING,usertype STRING,gender STRING,birthyear INT"
    ddlDataFrame = spark.read.schema(schema).csv(csvFile)
    print("**************************************")
    ddlDataFrame.show(n=10)
    print("************Print Schema DDL ******************")
    ddlDataFrame.printSchema()
    print("**************Number of records*****************")
    print("Number of records in ddl DataFrame = " + str(ddlDataFrame.count()))
    print("****************End of ddl schema*****************")

File: smeshram1/itmd-521/labs/week-04/scala/src/main/scala/assignment_01/assignment_01.scala
package main.scala.assignment_01

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

object assignment_01 {

  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder.appName("Assignment1").getOrCreate()

    if (args.length < 1) {
      println("Usage: DivvySet <Divvy_file_dataset>")
      spark.stop()
      sys.exit(1)
    }

    val data_source = args(0)

    // Read CSV with inferred schema
    val infer_divvy_df = spark.read.option("header", "true").option("inferSchema", "true").csv(data_source)
    printDataFrameInfo("Schema Inferred Data Frame:", infer_divvy_df)

    // Define StructType schema
    val struct_schema = StructType(
      Seq("trip_id", "bikeid", "tripduration", "from_station_id", "to_station_id", "birthyear")
        .map(field => StructField(field, IntegerType)) ++
        Seq("starttime", "stoptime", "from_station_name", "to_station_name", "usertype", "gender")
          .map(field => StructField(field, StringType))
    )

    // Read CSV with StructType schema
    val struct_divvy_df = spark.read.schema(struct_schema).option("header", "true").csv(data_source)
    printDataFrameInfo("Programmatically Defined Schema:", struct_divvy_df)

    // Define DDL schema
    val ddl_schema = "trip_id INT, starttime STRING, stoptime STRING, bikeid INT, tripduration INT, from_station_id INT, from_station_name STRING, to_station_id INT, to_station_name STRING, usertype STRING, gender STRING, birthyear INT"

    // Read CSV with DDL schema
    val ddl_df = spark.read.schema(ddl_schema).option("header", "true").csv(data_source)
    printDataFrameInfo("DDL Schema Data Frame:", ddl_df)

    // Filter and group by gender
    val female_to_station_df = infer_divvy_df.filter("gender = 'Female'").groupBy("to_station_name").count()
    printDataFrameInfo("Female Gender Data Frame grouped by to_station_name:", female_to_station_df, Some(10))

    spark.stop()
  }

  // Helper function to print DataFrame information
  def printDataFrameInfo(title: String, df: org.apache.spark.sql.DataFrame, numRows: Option[Int] = None): Unit = {
    println(title)
    df.printSchema()
    println(s"Count: ${df.count()}")
    numRows.foreach(n => df.show(n))
    println()
  }
}